### PR TITLE
Hitoashisaki

### PR DESCRIPTION
**Pull request by KorewaWatchful**
### Precomputed functions woooooo
- theyre functions, theyre precomputed
- @1whatleytay is an epic gamer who wrote most of this code

## images more to come stay tuned 
image

### CODE REVIEW

File: vita3k/app/src/screen_render.cpp [hunk 1]
```diff
@@ -134,6 +134,10 @@ void gl_screen_renderer::render(const HostState &host) {
         glEnableVertexAttribArray(uvAttrib);
 
         glBindTexture(GL_TEXTURE_2D, m_screen_texture);
+        if (display.base.address() < 10) {
+            LOG_WARN("Display base is {}", log_hex(display.base.address()));
+            return;
+        }
         const auto pixels = display.base.cast<void>().get(mem);
```
Comment (1whatleytay) [inline vita3k/app/src/screen_render.cpp:141]: Oh dear lord we can drop this

File: vita3k/gxm/include/gxm/types.h [hunk 2]
```diff
@@ -1440,3 +1448,40 @@ enum class GenericParameterType {
...
+    SCE_GXM_PRECOMPUTED_VERTEX_STATE_WORD_COUNT = 7,
+    SCE_GXM_PRECOMPUTED_FRAGMENT_STATE_WORD_COUNT = 9,
+    SCE_GXM_PRECOMPUTED_DRAW_WORD_COUNT = 11,
+};
+
+struct SceGxmPrecomputedDraw {
+    Ptr<const SceGxmVertexProgram> program;
+    Ptr<void> extra_data;
+
+	SceGxmPrimitiveType type;
```
Comment (1whatleytay) [inline vita3k/gxm/include/gxm/types.h:1462]: tabbing issues

File: vita3k/modules/SceGxm/SceGxm.cpp [hunk 3]
```diff
@@ -636,15 +733,21 @@ EXPORT(int, sceGxmGetParameterBufferThreshold) {
...
 EXPORT(int, sceGxmGetPrecomputedDrawSize) {
-    return UNIMPLEMENTED();
+    STUBBED("handle");
+
+    return 4;
 }
 
 EXPORT(int, sceGxmGetPrecomputedFragmentStateSize) {
-    return UNIMPLEMENTED();
+    STUBBED("handle");
```
Comment (1whatleytay) [inline vita3k/modules/SceGxm/SceGxm.cpp:742]: technically not stubbed anymore

File: vita3k/modules/SceGxm/SceGxm.cpp [hunk 4]
```diff
@@ -2255,7 +2404,7 @@ EXPORT(int, sceGxmUnmapFragmentUsseMemory, void *base) {
 }
 
 EXPORT(int, sceGxmUnmapMemory, void *base) {
-    assert(base != nullptr);
+    //assert(base != nullptr);
```
Comment (1whatleytay) [inline vita3k/modules/SceGxm/SceGxm.cpp:2408]: Why comment this out? Maybe replace it with a ret-error?

File: vita3k/modules/SceNgsUser/SceNgs.cpp [hunk 5]
```diff
@@ -238,7 +238,8 @@ EXPORT(int, sceNgsVoiceLockParams, std::uint32_t voice, std::uint32_t unk1, std:
     buffer_info->data = alloc(host.mem, 10, "SceNgs buffer stub");
     buffer_info->size = 10;
 
-    return STUBBED("Ngs buffer stubbed");
+	return -1;
+    //return STUBBED("Ngs buffer stubbed");
 }
```
Comment (1whatleytay) [inline vita3k/modules/SceNgsUser/SceNgs.cpp:243]: Ahh is this me? You can forget about it. I was debugging the p4g video when i stumbled across the persona4kindaworks branch and there are some debugging artifacts left.

**Comment by IllusionMan1212** [comment]
idk what any of this code means, but it's based and redpilled

File: vita3k/modules/SceSysmem/SceSysmem.cpp [hunk 6]
```diff
@@ -101,7 +101,7 @@ EXPORT(int, sceKernelFreeMemBlock, SceUID uid) {
 
     KernelState *const state = &host.kernel;
     const Blocks::const_iterator block = state->blocks.find(uid);
-    assert(block != state->blocks.end());
+    if (block == state->blocks.end()) return 0;
```
Comment (1whatleytay) [inline vita3k/modules/SceSysmem/SceSysmem.cpp:104]:
```suggestion
    if (block == state->blocks.end())
        return 0;
```

When does this happen?

**Comment by KorewaWatchful** [comment]
thank you

File: vita3k/shader/src/spirv_recompiler.cpp [hunk 7]
```diff
@@ -699,7 +699,7 @@ static SpirvShaderParameters create_parameters(spv::Builder &b, const SceGxmProg
 
     spv::Id f32_v4_type = b.makeVectorType(f32_type, 4);
     spv::Id pa_arr_type = b.makeArrayType(f32_v4_type, b.makeIntConstant(32), 0);
-    spv::Id sa_arr_type = b.makeArrayType(f32_v4_type, b.makeIntConstant(32), 0);
+    spv::Id sa_arr_type = b.makeArrayType(f32_v4_type, b.makeIntConstant(40), 0);
```
Comment (1whatleytay) [inline vita3k/shader/src/spirv_recompiler.cpp:702]: This is a sad workaround. No clue how we get sa indices higher than 128, probably smart to look into.

File: vita3k/modules/SceNgsUser/SceNgs.cpp (hunk 5 above)
Comment (KorewaWatchful) [reply vita3k/modules/SceNgsUser/SceNgs.cpp:243]: forgot to remove that back yikes

File: vita3k/modules/SceSysmem/SceSysmem.cpp (hunk 6 above)
Comment (KorewaWatchful) [reply vita3k/modules/SceSysmem/SceSysmem.cpp:104]: i would like to ask you that question xD

File: vita3k/modules/SceGxm/SceGxm.cpp (hunk 3 above)
Comment (KorewaWatchful) [reply vita3k/modules/SceGxm/SceGxm.cpp:742]: should i remove the stubs

File: vita3k/gxm/include/gxm/types.h (hunk 2 above)
Comment (KorewaWatchful) [reply vita3k/gxm/include/gxm/types.h:1462]: ill run format visual studio hurts my brain

File: vita3k/app/src/screen_render.cpp (hunk 1 above)
Comment (KorewaWatchful) [reply vita3k/app/src/screen_render.cpp:141]: roger

**Comment by Zangetsu38** [comment]
Very lot format error and all code is set by desgroup 🤔 not understund here ^^

**Comment by KorewaWatchful** [comment]
Believe it or not, there's actually some code setup by me.

**Comment by Zangetsu38** [comment]
> Believe it or not, there's actually some code setup by me.

Ah yes i now that, i want said, pr is wrong ^^ more needed, hmm using commit by desgroup and after set you proper commit 😊
I have already used commit by other and after set my proper commit in second is easy

**Comment by KorewaWatchful** [comment]
too much hassle and would have a lot of commits, this way desgroup is co-author and everything is nice and working with the emulator instead of the old commit.

**Comment by Zangetsu38** [comment]
> too much hassle and would have a lot of commits, this way desgroup is co-author and everything is nice and working with the emulator instead of the old commit. 

No here is just set you like author, and use like i said, i have exp,ainsi you you commit desgroup code with set author desgroup, and after set your proper code on your own commit 😊 is very easy set this.
But good jof on your code works

**Comment by 1whatleytay** [comment]
To be fair, I'm not sure if we're going to merge a commit called "precomputed winning" XD.

I think two commits is okay (like one for the gxm functions and another for the shader configuring). I'm okay with one commit though.

File: vita3k/modules/SceGxm/SceGxm.cpp (hunk 3 above)
Comment (1whatleytay) [reply vita3k/modules/SceGxm/SceGxm.cpp:742]: Why not. Set the other two sizes to 0 btw.

File: vita3k/modules/SceSysmem/SceSysmem.cpp (hunk 6 above)
Comment (1whatleytay) [reply vita3k/modules/SceSysmem/SceSysmem.cpp]: Yeah we can ignore this probably XD

**Comment by Ristellise** [comment]
Looks intresting... Though I still dont really understand it all XD

**Comment by 1whatleytay** [comment]
What needs to be done for this to be merged?

**Comment by KorewaWatchful** [comment]
I need to not be grounded and get my pc back so we can finish fixing the texture bug that causes the crashes.

**Comment by 1whatleytay** [comment]
Damn, grounded and quarantined? That sucks...

I'm okay to merge if the texture bug isn't doesn't cause a regression. Will wait until you're ready to debug though :D

**Review by sunho — APPROVED** [review]

**Pull request merged**